Refactor project to be able to integrate webpack and support IE11

Spring3 opened the pull request:
<!-- Provide a general summary of the pull request changes in the title above -->
<!-- Replace the rather silly content below with your acutal pull request information -->

## Proposed Changes
  Whoosh... there's a lot..
  - Integrated webpack with babel
  - To support IE 10+, I had to add core-js polyfill for `Object.entries`
  - Refactored the project to not load static items in runtime
  - Restructured the project to make it look more clean and understandable
  - Got rid of async/await
  - instead of `HTMLElement.remove()`, I used `HTMLElement.parentNode.removeChild(e)` which is supported by IE and not so modern browsers
## Additonal Comments
* Previously you tried to detect the folder, from which the static files were served (using `document.currentScript`, which is not supported in IE at all and [can not be safely polyfilled](https://stackoverflow.com/questions/403967/how-may-i-reference-the-script-tag-that-loaded-the-currently-executing-script) as we don't know if the script will be loaded async or sync

* I removed `__loadCSS` function and `__loadTemplate` (ish) functions that tried to fetch the static files in runtime and instead am using webpack to inject html template as a string directly in js bundle, and css is available as a separate bundle that the user needs to place on the page separately

* In my opinion, you overused the modern features for such a simple module, so using async, await would lead to either putting a `regenerator-runtime` module that weights 24kb and would bloat this module, or converting async await to Promise instead and then use the promise polyfill that is slightly more lightweight. Fortunately, when I reviewed the code, I noticed that you used `fetch` only to get the static files in runtime, so I came up with a way to get rid of any async code and just have the static files loaded once and then reused.

All the static files are now placed in the `dist` folder. To get it, firstly run `npm i` to download the required packages and then build the project with `npm run build`. Js bundle will be minified, css bundle will be available in both variations and other files like fonts and audio will be copied untouched. 

~~A small notice about the demo website. `index.html` file in the project root is now used as a template, meaning that it does not have any links to the bundles or source files. If you want your demo website to work, you need to serve the page, that is generated in the `dist` folder instead.~~

* In order to use this module, you need to copy the `dist` folder and serve it with the rest of your static files

The last point is that if I were you, I would rewrite the main js bundle file without using any es6 features, but instead making it rely on closure. There is no point in using es6 classes here in my opinion. But I will let you decide of course. Also, consider using some "default" fonts as having some custom ones is of course great, but this means additional payload and work to do when trying to integrate this module. Either use the defaults, and let the user set the font of their choice, or keep 1 font weight and remove the others, because you have `Black, Bold, Heavy, Light, Medium, Regular, Semibold, Thin, Ultrathin` - don't you think that's too many? :) Especially since each one of them weights 200+KB

Good luck and if you have any questions - feel free to ask 

resolves #7 

Oh, and here's the proof of work in IE 11. 

![oct-05-2018 06-02-16](https://user-images.githubusercontent.com/4171202/46514050-4550ad00-c864-11e8-9519-499471a2abe7.gif)


And here it works in IE 10
![oct-05-2018 06-05-47](https://user-images.githubusercontent.com/4171202/46514148-cb6cf380-c864-11e8-90a3-fbb06a16be0d.gif)

Spring3 committed:
Refactor project to be able to interact with webpack and support IE11

MattIPv4 commented on src/macOSNotif.js:
```diff
@@ -336,7 +310,7 @@ class macOSNotifJS {
         // Remove fully once animation completed
         setTimeout(() => {
-            this.container.parentElement.remove();
+            this.container.parentElement.removeChild(this.container);
```
Should this not be `this.container.parentElement.parentElement.removeChild(this.container.parentElement);` to remove the parent on container (which is what is needed)?

Spring3 replied:
indeed, good catch

MattIPv4 commented on index.html:
```diff
@@ -21,7 +21,6 @@
 <head>
     <meta charset="UTF-8">
     <title>macOSNotifJS</title>
-    <link href="plugin/res/macOSNotif.fonts.min.css" rel="stylesheet" type="text/css"/>
```
This doesn't appear to have been replaced?

MattIPv4 replied:
Just noticed your comment about this, is there a way to get this working or not? On the package site it serves this page, so having it as a working demo would be nice.

MattIPv4 commented on dist/index.html:
```diff
@@ -0,0 +1,165 @@
+            margin-left: 0.5em;
+        }
+    </style>
+<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="macOSNotif.css" rel="stylesheet"><script type="text/javascript" src="macOSNotif.min.js"></script></head>
```
Can this be tidied to have new lines and conform to XHTML with self-closing slashes?
Script should also be at the end of the body not in the head?

MattIPv4 commented on index.html:
```diff
@@ -144,9 +143,8 @@ <h3>Demos:</h3>
 </button>
 <code>macOSNotif({sounds:true})</code><br/>
 
-<script src="plugin/macOSNotif.js"></script>
```
This doesn't appear to have been replaced?

MattIPv4 replied:
Just noticed your comment about this, is there a way to get this working or not? On the package site it serves this page, so having it as a working demo would be nice.

MattIPv4 commented on package.json:
```diff
@@ -0,0 +1,40 @@
+  },
+  "keywords": [],
+  "author": "MattIPv4",
+  "license": "ISC",
```
Should be `AGPL-3.0-only`?

Spring3 replied:
Yes, you are right. Didn't notice it (was auto generated by npm). I will update this PR later today

MattIPv4 reviewed (changes requested):
That's my two cents for now. I'll clone it later (probably tomorrow) to test it fully in the new state.

**Thank you for all the work though, this is something I would have never done :)**

Spring3 commented:
Regarding the html - I will see what I can do

Spring3 committed:
Fix license type in package.json and container removal

Spring3 committed:
Include audio files in js bundle. Fix index.html layout

Spring3 commented:
@MattIPv4 I added 2 commits:
1) the first one is minor and it just changes the type of license in `package.json` as well as removes the notification container properly
2) the second one contains a bit bigger changes:
- audio files are now included in the js bundle as base64 encoded strings.
How so?
Since audio files are required in the js file, they are required by the following path: `require('./audio/<file>.<ext>')`. The html, that is outside the dist folder, references the script by the followint path: `src="/dist/macOSNotif.min.js"`. When the script file is processed, it instructs the browser to fetch the audio files from `./audio/` folder. If our html page was inside the `dist` folder, it would work perfectly, but right now as it is in the root of the project, there is no `audio` folder there, so audio files resulted in 404 response. Having them base64 encoded brought the following consequences:
### Advantages:
1) the user does not have to maintain the folder structure anymore. If I let's say decided to rename the `audio` folder to `sounds`, it would break previously, but with the current approach, it will be completely fine.
2) No need to make additional requests to get the sound files and at the same moment, no need to serve them anymore.
### Disadvantages:
1) The size of the bundle is now += the sum of the sizes of the audio files (+40kb). A possible fix here would be to either gzip the bundle when you serve it or not use the audio files (especially since I think that it would not play on mobile devices (but not sure, I haven't checked) )
- The last change was that I fixed he structure of the generated index.html page. Now it's working from the root of the project and has a cleaner structure inside

MattIPv4 commented:
Morning, this looks very promising now.
One more thing that I can think of currently, why is there an un-minified css file in the dist folder?

Spring3 committed:
Generate only min.css bundle

Spring3 commented:
@MattIPv4 Evening, I'm glad to hear that you find the implementation promising.

I have added another commit to generate only the min.css bundle. Please take a look when you have time

MattIPv4 commented on index.html:
```diff
@@ -144,9 +152,8 @@ <h3>Demos:</h3>
-<script src="plugin/macOSNotif.js"></script>
 <script>
-    if(window.location.hostname != "macosnotif.js.org") window.location.replace("https://macosnotif.js.org");
+    // if(window.location.hostname != "macosnotif.js.org") window.location.replace("https://macosnotif.js.org");
```
Is there a reason for this being commented out now?
_I presume this was done for testing?_

MattIPv4 commented on index.tpl:
```diff
@@ -0,0 +1,175 @@
+<!--
```
I presume this is the template used to generate index.html?

Spring3 replied:
Exactly. If you take a closer look, you will see the insertion points.
For example
```
<meta name="viewport" content="<%= htmlWebpackPlugin.options.meta.viewport %>"/>
```
or
```
<% for (var css in htmlWebpackPlugin.files.css) { %>
    <link href="<%= htmlWebpackPlugin.files.css[css] %>" rel="stylesheet"/>
<% } %>
``` 
to inject the css

MattIPv4 commented on src/macOSNotif.js:
```diff
@@ -446,7 +419,7 @@ class macOSNotifJS {
 }
 
 // eslint-disable-next-line no-unused-vars
```
As the below is now set as an attribute of window, this eslint line shouldn't be needed?

Spring3 commented:
Fixed (uncommented the redirect in .html file and also removed the eslint rule ignore in js file)

Spring3 commented:
And FYI, I do not have rights to merge the PR, so once you validate and confirm that it does not introduce any regressions, feel free to merge :)

MattIPv4 commented:
Yeah I'll clone this now and verify. Once I'm happy I'll merge :)

MattIPv4 commented:
Okay just cloned and looks good, I've only tested the index cause if that works I'm happy the plugin works as a whole.

~~One tiny thing, evidently a bug in my version prior to this as well. In the demo page (index), the sound notification triggered at page load doesn't play its sound for me in chrome. Pressing the button later manually plays the sound fine. Any ideas why?~~ Ignore this, I had missed the error in console. Chrome no longer lets you play sounds without user interaction first.

MattIPv4 reviewed (approved):
Final approval, good to go.
